Microsoft.Toolkit/Observables/ReadOnlyObservableGroupedCollection.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.Toolkit.Observables.Collections
+{
+    /// <summary>
+    /// A read-only list of groups.
+    /// </summary>
+    /// <typeparam name="TKey">The type of the group key.</typeparam>
+    /// <typeparam name = "TValue" > The type of the items in the collection.</typeparam>
+    public sealed class ReadOnlyObservableGroupedCollection<TKey, TValue> :
```

Comment:
I'm probably missing something here, but why does this class need to reimplement everything from scratch instead of just being defined like `ObservableGroupedCollection<TKey, TValue>`, but with the `ReadOnly{...}` variant? As in:

```csharp
public sealed class ReadOnlyObservableGroupedCollection<TKey, TValue> :
    ReadOnlyObservableCollection<ReadOnlyObservableGroup<TKey, TValue>>
{
    // ...
}
```

I'm a bit confused 🤔

Reply:
It is mostly because I need to convert the `ObservableGroup<TKey, TValue>` to their *read-only* counter parts before raising the `INotificationCollectionChanged` events. 

`ReadOnlyObservableGroupedCollection<TKey, TValue>` is a *read-only* list of *read-only* groups which I'm building from a *mutable* list of *mutable* groups.

Unfortunately, the system `ReadOnlyObservableCollection<T>` does not allow us to convert the `T` to a `<read-only> T` (at least I haven't found how to do it 🙂).

Reply:
I'm really not a big fan of having all that code being duplicated there (I know you're just trying to work around that limitation). I think I have found a way to still avoid having all that code there, by adding this 3rd constructor to `ReadOnlyObservableGroup<TKey, TValue>`:

```csharp
/// <summary>
/// Initializes a new instance of the <see cref="ReadOnlyObservableGroup{TKey, TValue}"/> class.
/// </summary>
/// <param name="key">The key of the group.</param>
/// <param name="collection">The collection of items to add in the group.</param>
public ReadOnlyObservableGroup(TKey key, IEnumerable<TValue> collection)
    : base(new ObservableCollection<TValue>(collection))
{
    Key = key;
}
```

Which then lets us define `ReadOnlyGroupedObservableCollection<TKey, TValue>` like this:

```csharp
/// <summary>
/// An observable list of observable groups.
/// </summary>
/// <typeparam name="TKey">The type of the group key.</typeparam>
/// <typeparam name="TValue">The type of the items in the collection.</typeparam>
public sealed class ReadOnlyObservableGroupedCollection<TKey, TValue> : ReadOnlyObservableCollection<ReadOnlyObservableGroup<TKey, TValue>>
{
    /// <summary>
    /// Initializes a new instance of the <see cref="ReadOnlyObservableGroupedCollection{TKey, TValue}"/> class.
    /// </summary>
    /// <param name="collection">The source collection to wrap.</param>
    public ReadOnlyObservableGroupedCollection(ObservableGroupedCollection<TKey, TValue> collection)
        : this(collection.Select(g => new ReadOnlyObservableGroup<TKey, TValue>(g)))
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ReadOnlyObservableGroupedCollection{TKey, TValue}"/> class.
    /// </summary>
    /// <param name="collection">The initial data to add in the grouped collection.</param>
    public ReadOnlyObservableGroupedCollection(IEnumerable<ReadOnlyObservableGroup<TKey, TValue>> collection)
        : base(new ObservableCollection<ReadOnlyObservableGroup<TKey, TValue>>(collection))
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ReadOnlyObservableGroupedCollection{TKey, TValue}"/> class.
    /// </summary>
    /// <param name="collection">The initial data to add in the grouped collection.</param>
    public ReadOnlyObservableGroupedCollection(IEnumerable<IGrouping<TKey, TValue>> collection)
        : this(collection.Select(g => new ReadOnlyObservableGroup<TKey, TValue>(g.Key, g)))
    {
    }
}
```

With just some constructor magic, and no code duplication at all 😄
The exposed API surface is exactly the same, or rather, this version also has two more public constructors, which might come in handy to some users. What do you think?

Reply:
Unresolving to not forget it. 

I'll give it a try. My concern would be the `CollectionChanged` event args. It seems that averriding [ObservableCollection<T>.OnCollectionChanged(NotifyCollectionChangedEventArgs)](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.observablecollection-1.oncollectionchanged?view=netframework-4.8#System_Collections_ObjectModel_ObservableCollection_1_OnCollectionChanged_System_Collections_Specialized_NotifyCollectionChangedEventArgs_) is a good candidate to make it work.

Reply:
This should actually handle that automatically without the need to override anything, as the new _readonly_ groups are passed directly to the base constructor, so the collection never actually sees the _mutable_ versions at all. Let me know how it goes! 😄